tests/load/gcp/prom-query-locust.tpl.conf

```diff
@@ -0,0 +1,37 @@
+#cloud-config
+
+write_files:
+- path: /etc/systemd/system/locust.service
+  content: |
+    [Unit]
+    Description=prom query locust
+    After=network.target
+
+    [Service]
+    Type=simple
+    Restart=always
+    ExecStart=python3 -m locust -f /home/ubuntu/prom-query-locustfile.py --host ${PROM_URL} --users ${USERS} --spawn-rate 10 --headless
+
+    #[Install]
+    #WantedBy=multi-user.target
+
+package_update: true
+
+packages:
+- python3-pip
+- jq
+
+runcmd:
+- python3 -m pip install locust
+
+- |
+  # wait until the lma node is up
+  until curl -s --connect-timeout 2.0 ${PROM_URL}/api/v1/targets; do sleep 5; done
+  
+  # without piping to jq, curl would return successfully with "nginx 404 not found"
+  until [ $(sh -c "curl -s --connect-timeout 2.0 ${PROM_URL}/api/v1/targets | jq '.data.activeTargets[].health' 2>/dev/null | wc -l") -gt 1 ]; do sleep 5; done
```

Comment:
As above

Reply:
I would need some bash help with this one:
Just adding timeout,
```bash
timeout 600 bash -c "until [ $(sh -c 'curl -s --connect-timeout 2 --max-time 5 http://pd-ssd-4cpu-16gb.us-central1-a.c.lma-light-load-testing.internal/prom/api/v1/targets | jq ".data.activeTargets[].health" 2>/dev/null | wc -l') -gt 1 ]; do sleep 5; done"
```
doesn't work because the variable is evaluated only once, so at startup, this line becomes
```bash
+ timeout 600 bash -c until [ 0 -gt 1 ]; do sleep 5; done
```
which timeouts and never starts locust, and later on when I manually run `/var/lib/cloud/instance/scripts/runcmd`:
```bash
+ timeout 600 bash -c until [ 6 -gt 1 ]; do sleep 5; done
```
returns immediately and starts locust.

How should I modify this so that the expression is evaluated every loop iteration?

Reply:
Two things: 

- Currently you're sort of needlessly start `sh` within your `bash` shell
- You just need to escape the expression you pass to `bash -c` to ensure it's expanded by the bash instance itself

I just setup a simple web server that returns similar data, and this seemed to work fine:

```bash
timeout 600 bash -c "until [[ \$(curl -s --connect-timeout 2 --max-time 5 http://pd-ssd-4cpu-16gb.us-central1-a.c.lma-light-load-testing.internal/prom/api/v1/targets | jq '.data.activeTargets[].health' | wc -l) -gt 1 ]]; do sleep 1; done"
```

See how the `$` is escaped for the `$()`? :)

Reply:
Or for bonus points on readability, replace with this:

```bash
timeout 10 bash << 'EOF'
set -euxo pipefail

_num_targets() {
  URL=http://localhost:5000/test
  echo $(curl -s --connect-timeout 2 --max-time 5 $URL | jq '.data.activeTargets[].health' | wc -l)
}

until [[ $(_num_targets) -gt 1 ]]; do sleep 1; done
echo "SUCCESS"
EOF
```

Note how the `'EOF'` is quoted at the start, which prevents variable expansion so you don't need to escape the `$()`

Reply:
I mentioned this two days ago in chat, and I really do think that Python one-liners quickly become more unreadable than even Perl one-liners did, it's an option.

*SHELL SCRIPTING CONSIDERED HARMFUL*. No, really, Perl got a lot of popularity because Perl5 was on pretty much every system, and it behaved the same everywhere. Even bash does not behave the same depending on the version, and it has lots of ugly edge cases like this because it's unintuitive.

`until` is a special case of `while`, to the point that they're literally in the same `case:` stanzas in the Bash source.

The answer, though, is something more like:

```bash
TARGETS=$(curl -s --connect-timeout 2.0 ${PROM_URL}/api/v1/targets | jq '.data.activeTargets[].health' 2>/dev/null | wc -l"); until [ $TARGETS -gt 1 ]; do TARGETS=$(curl -s --connect-timeout 2.0 ${PROM_URL}/api/v1/targets | jq '.data.activeTargets[].health' 2>/dev/null | wc -l"); sleep 5; done
```

If you can live with the extra 5 seconds, initialize `TARGETS` to zero so the line is shorter.

Or you can pipe the whole thing with something like:
```bash
until < $( ...)
```

But really, I suspect that it's some additional cruft, and that cutting out the additional subshell of `sh -c` would work:
```bash
timeout 600 bash -c "until [ $(curl -s --connect-timeout 2 --max-time 5 http://pd-ssd-4cpu-16gb.us-central1-a.c.lma-light-load-testing.internal/prom/api/v1/targets | jq ".data.activeTargets[].health" 2>/dev/null | wc -l) -gt 1 ]; do sleep 5; done"
```

Reply:
Thanks @rbarry82, that is indeed a neat solution. Going with the heredoc though for readability.